// gspeak/dist/src/gSpeak.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.gSpeak = void 0;
const fs_1 = require("fs");
const stream_1 = require("stream");
const gToken_1 = __importDefault(require("./gToken"));
const languages_1 = __importDefault(require("./languages"));
const GOOGLE_TTS_URL = 'https://translate.google.com/translate_tts';
const MAX_CHARS = 100;
const escapeRegExp = (s) => s.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&');
class gSpeak {
    lang;
    text;
    text_parts;
    debug;
    constructor(text, lang = 'en', debug = false) {
        this.debug = debug;
        if (!languages_1.default[lang.toLowerCase()])
            throw new Error(`Language not supported: ${lang}`);
        this.lang = lang.toLowerCase();
        if (!text)
            throw new Error('No text to speak');
        this.text = text;
        let parts = text.length <= MAX_CHARS ? [text] : this._tokenize(text, MAX_CHARS);
        parts = parts.map(p => p.replace(/\\n/g, '').trim()).filter(p => p.length > 0);
        this.text_parts = parts;
    }
    getHeaders() {
        return {
            'Referer': 'http://translate.google.com/',
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36'
        };
    }
    getUrl(part, idx) {
        const params = new URLSearchParams({
            ie: 'UTF-8',
            q: part,
            tl: this.lang,
            total: String(this.text_parts.length),
            idx: String(idx),
            client: 'tw-ob',
            textlen: String(part.length),
            tk: (0, gToken_1.default)(part)
        });
        return `${GOOGLE_TTS_URL}?${params}`;
    }
    stream() {
        const pass = new stream_1.PassThrough();
        (async () => {
            for (const [idx, part] of this.text_parts.entries()) {
                const url = this.getUrl(part, idx);
                if (this.debug)
                    console.log({ part, idx, url });
                const res = await fetch(url, { headers: this.getHeaders() });
                if (!res.ok)
                    throw new Error(`Request failed: ${res.status}`);
                pass.write(Buffer.from(await res.arrayBuffer()));
            }
            pass.end();
        })().catch(err => pass.destroy(err));
        return pass;
    }
    save(filePath, callback) {
        ;
        (async () => {
            for (const [idx, part] of this.text_parts.entries()) {
                const url = this.getUrl(part, idx);
                if (this.debug)
                    console.log({ part, idx, url });
                const res = await fetch(url, { headers: this.getHeaders() });
                if (!res.ok)
                    throw new Error(`Request failed: ${res.status}`);
                await new Promise(async (resolve, reject) => {
                    const ws = (0, fs_1.createWriteStream)(filePath, { flags: idx > 0 ? 'a' : 'w' });
                    ws.on('finish', resolve);
                    ws.on('error', reject);
                    ws.end(Buffer.from(await res.arrayBuffer()));
                });
            }
        })().then(() => callback(null)).catch(callback);
    }
    _tokenize(text, max_size) {
        const punc = '¡!()[]¿?.,;:—«»\n'.split('').map(escapeRegExp);
        const parts = text.split(new RegExp(punc.join('|')));
        return parts.flatMap(p => this._minimize(p, ' ', max_size));
    }
    _minimize(str, delim, max_size) {
        if (str.length <= max_size)
            return [str];
        const idx = str.lastIndexOf(delim);
        return [str.substring(0, idx), ...this._minimize(str.substring(idx), delim, max_size)];
    }
    static get languages() {
        return languages_1.default;
    }
}
exports.gSpeak = gSpeak;
exports.default = gSpeak;
