// stickers-formatter/dist/internal/Metadata/StickerMetadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Utils_1 = __importDefault(require("../../Utils"));
class StickerMetadata {
    pack;
    author;
    categories;
    id;
    crop = false;
    full = false;
    constructor(pack = '', author = '', categories = [], id = Utils_1.default.generateStickerID()) {
        this.pack = pack;
        this.author = author;
        this.categories = categories;
        this.id = id;
    }
    static from = (object) => {
        return new StickerMetadata(object.pack, object.author, object.categories, object.id);
    };
    setPack = (title) => {
        this.pack = title;
        return this;
    };
    setAuthor = (author) => {
        this.author = author;
        return this;
    };
    setId = (id) => {
        this.id = id;
        return this;
    };
    setCrop = (value) => {
        this.crop = value;
        this.full = !value;
        return this;
    };
    setFull = (value) => {
        this.crop = !value;
        this.full = value;
        return this;
    };
    setCategories = (categories) => {
        this.categories = (typeof categories === 'string' ? categories.split(',').map((emoji) => emoji.trim()) : categories);
        return this;
    };
    toJSON = () => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const obj = {};
        Object.keys(this)
            .filter((key) => typeof this[key] !== 'function')
            .forEach((key) => (obj[key] = this[key]));
        return obj;
    };
}
exports.default = StickerMetadata;
